examples/sde_gan.py

```diff
@@ -0,0 +1,523 @@
+        xs = xs.transpose(0, 1)  # switch t_size and batch_size
+        ys = self._readout(xs)
+
+        ###################
+        # Normalise the data to the form that the discriminator expects, in particular including time as a channel.
+        ###################
+        t_size = ts.size(0)
+        ts = ts.unsqueeze(0).unsqueeze(-1).expand(batch_size, t_size, 1)
+        return torchcde.linear_interpolation_coeffs(torch.cat([ts, ys], dim=2))
+
+
+###################
+# Next the discriminator. Here, we're going to use a neural controlled differential equation (neural CDE) as the
+# discriminator, just as in the "Neural SDEs Made Easy: SDEs are Infinite-Dimensional GANs" paper.
+#
+# There's actually a few different (roughly equivalent) ways of making the discriminator work. The curious reader is
+# strongly encouraged to have a read of the comment at the bottom of this file for an in-depth explanation.
+###################
+class DiscriminatorFunc(torch.nn.Module):
+    def __init__(self, data_size, hidden_size, mlp_size, num_layers):
+        super(DiscriminatorFunc, self).__init__()
+        self._data_size = data_size
+        self._hidden_size = hidden_size
+        # tanh is important for model performance
+        self._module = MLP(1 + hidden_size, hidden_size * (1 + data_size), mlp_size, num_layers, tanh=True)
+
+    def forward(self, t, h):
+        # t has shape ()
+        # h has shape (batch_size, hidden_size)
+        t = t.expand(h.size(0), 1)
+        th = torch.cat([t, h], dim=1)
+        return self._module(th).view(h.size(0), self._hidden_size, 1 + self._data_size)
+
+
+class Discriminator(torch.nn.Module):
+    def __init__(self, data_size, hidden_size, mlp_size, num_layers):
+        super(Discriminator, self).__init__()
+        self._initial = MLP(1 + data_size, hidden_size, mlp_size, num_layers, tanh=False)
+        self._func = DiscriminatorFunc(data_size, hidden_size, mlp_size, num_layers)
+        self._readout = torch.nn.Linear(hidden_size, 1)
+
+        # Final layer has an easier problem to solve, so increase its learning rate.
+        self._readout.weight.register_hook(lambda grad: 100 * grad)
+        self._readout.bias.register_hook(lambda grad: 100 * grad)
+
+    def forward(self, ys_coeffs):
+        # ys_coeffs has shape (batch_size, t_size, 1 + data_size)
+        # The +1 corresponds to time. When solving CDEs, It turns out to be most natural to treat time as just another
+        # channel: in particular this makes handling irregular data quite easy, when the times may be different between
+        # different samples in the batch.
+
+        Y = torchcde.LinearInterpolation(ys_coeffs)
+        Y0 = Y.evaluate(Y.interval[0])
+
+        # TODO: step size
+        h0 = self._initial(Y0)
+        # We happen to use midpoint with step_size=1e-1 for consistency with the SDE solve, but that isn't important.
```

Comment:
Does this refer to the step size for generating the data, or that for the generator solve?

Reply:
I've removed this now.